crates/fortitude_preprocessor/src/cpp.rs

```diff
@@ -0,0 +1,852 @@
+use crate::cpp_tokens::{CppDirectiveKind, CppTokenIterator, CppTokenKind};
+use anyhow::{Context, anyhow};
+use chrono::prelude::*;
+use lazy_regex::regex_captures;
+use ruff_source_file::{OneIndexed, SourceCode};
+use ruff_text_size::TextSize;
+use std::collections::HashMap;
+use std::path::{Path, PathBuf};
+
+/// Enum describing where a snippet came from.
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub enum Provenance {
+    /// A system-defined macro, e.g. __DATE__ or __TIME__.
+    SystemDefined,
+    /// A user-defined macro, e.g. passed in via command line.
+    UserDefined,
+    /// A macro defined in the source file.
+    FileDefined {
+        start: usize,
+        end: usize,
+        path: PathBuf,
+    },
+    /// Plain text from an included file.
+    IncludeText {
+        start: usize,
+        end: usize,
+        path: PathBuf,
+    },
+    /// Plain text in the source file, not from a macro.
+    LocalText { start: usize, end: usize },
+}
+
+/// A snippet of text with provenance information.
+pub struct Snippet {
+    text: String,
```

Comment:
Is this the transformed text? A possible future optimisation would be to use `Cow<str>` here to avoid unnecessary allocations

Reply:
Yes, the snippets will be the final text of the preprocessed file. I haven't used `Cow<str>` before, but I'd be happy to look into it if it'll save some copies.

Reply:
I tried this with `Cow<'a, str>` and ran into a host of lifetime issues. For now I've left it as a `String`, though I was able to use `Cow<'a, str>` in `LogicalLine`.